modules/module-postgres-storage/src/storage/batch/PostgresBucketBatch.ts

```diff
@@ -0,0 +1,891 @@
+        last_checkpoint = COALESCE(
+          ${{ type: 'int8', value: update.last_checkpoint }},
+          last_checkpoint
+        ),
+        last_checkpoint_ts = ${{ type: 1184, value: update.last_checkpoint_ts }},
+        last_checkpoint_lsn = ${{ type: 'varchar', value: update.last_checkpoint_lsn }}
+      WHERE
+        id = ${{ type: 'int4', value: this.group_id }}
+      RETURNING
+        id,
+        state,
+        last_checkpoint,
+        last_checkpoint_lsn
+    `
+      .decoded(StatefulCheckpoint)
+      .first();
+
+    await notifySyncRulesUpdate(this.db, doc!);
+
+    this.persisted_op = null;
+    this.last_checkpoint_lsn = lsn;
+    return true;
+  }
+
+  async keepalive(lsn: string): Promise<boolean> {
+    if (this.last_checkpoint_lsn != null && lsn <= this.last_checkpoint_lsn) {
+      // No-op
+      return false;
+    }
+
+    if (lsn < this.no_checkpoint_before_lsn) {
+      return false;
+    }
+
+    if (this.persisted_op != null) {
+      // The commit may have been skipped due to "no_checkpoint_before_lsn".
+      // Apply it now if relevant
+      logger.info(`Commit due to keepalive at ${lsn} / ${this.persisted_op}`);
+      return await this.commit(lsn);
+    }
+
+    const updated = await this.db.sql`
+      UPDATE sync_rules
+      SET
+        snapshot_done = ${{ type: 'bool', value: true }},
+        last_checkpoint_lsn = ${{ type: 'varchar', value: lsn }},
+        last_fatal_error = ${{ type: 'varchar', value: null }},
+        last_keepalive_ts = ${{ type: 1184, value: new Date().toISOString() }}
+      WHERE
+        id = ${{ type: 'int8', value: this.group_id }}
+      RETURNING
+        id,
+        state,
+        last_checkpoint,
+        last_checkpoint_lsn
+    `
+      .decoded(StatefulCheckpoint)
+      .first();
+
+    await notifySyncRulesUpdate(this.db, updated!);
+
+    this.last_checkpoint_lsn = lsn;
+    return true;
+  }
+
+  async markSnapshotDone(
+    tables: storage.SourceTable[],
+    no_checkpoint_before_lsn: string
+  ): Promise<storage.SourceTable[]> {
+    const ids = tables.map((table) => table.id.toString());
+
+    await this.db.transaction(async (db) => {
+      await db.sql`
+        UPDATE source_tables
+        SET
+          snapshot_done = ${{ type: 'bool', value: true }}
+        WHERE
+          id IN (
+            SELECT
+              (value ->> 0)::TEXT
+            FROM
+              jsonb_array_elements(${{ type: 'jsonb', value: ids }}) AS value
+          );
+      `.execute();
+
+      if (no_checkpoint_before_lsn > this.no_checkpoint_before_lsn) {
+        this.no_checkpoint_before_lsn = no_checkpoint_before_lsn;
+
+        await db.sql`
+          UPDATE sync_rules
+          SET
+            no_checkpoint_before = ${{ type: 'varchar', value: no_checkpoint_before_lsn }},
+            last_keepalive_ts = ${{ type: 1184, value: new Date().toISOString() }}
+          WHERE
+            id = ${{ type: 'int8', value: this.group_id }}
```

Comment:
There's currently a mix of `int8` and `int4` used for this id in code - I think just stick to `int4`.

Reply:
Thanks. Good catch. This one got a bit confusing since pgwire would return `bigint` type values for `integer` columns. I've changed this to `int4` with the relevant `number` Typescript types. I've used a codec to workaround pgwire returning a bigint.